### PR TITLE
Fix/limit thumbnail size

### PR DESCRIPTION
**Pull request by lpoli**
### Changes
Delete object function has been optimized by removal redundant of logic and addition of goroutine for disk operations.

### Fixes
Thumbnail file size is limited to 1MB

### CODE REVIEW

File: code/go/0chain.net/blobbercore/allocation/deletefilechange.go [hunk 1]
```diff
@@ -58,20 +56,64 @@ func (nf *DeleteFileChange) DeleteTempFile() error {
...
+					Where("allocation_id=? AND content_hash=?", nf.AllocationID, res.ContentHash).
+					Count(&count)
+
+				go func(res Result, count int64) {
+					defer func() {
+						<-limitCh
+						wg.Done()
+					}()
+
+					if count == 0 {
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/allocation/deletefilechange.go]: We can do this count checking out of the gorountine, so if it's not 0, we don't need to start the goroutine.

File: code/go/0chain.net/blobbercore/handler/file_command_update.go [hunk 2]
```diff
@@ -54,6 +58,14 @@ func (cmd *UpdateFileCommand) IsValidated(ctx context.Context, req *http.Request
 		return common.NewError("invalid_operation", "Operation needs to be performed by the owner, collaborator or the payer of the allocation")
 	}
 
+	_, thumbHeader, _ := req.FormFile(UploadThumbnailFile)
+	if thumbHeader != nil {
+		if thumbHeader.Size > MaxThumbnailSize {
+			return common.NewError("max_thumbnail_size",
+				fmt.Sprintf("Thumbnail size %d should not be greater than %d", thumbHeader.Size, MaxThumbnailSize))
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/handler/file_command_update.go]: We'd better not capitalize error messages, as they are usually printed following other context.

File: code/go/0chain.net/blobbercore/reference/object.go [hunk 3]
```diff
@@ -3,77 +3,48 @@ package reference
...
-		found := false
-		for _, child := range dirRef.Children {
-			if child.Name == tSubDirs[treelevel] && child.Type == DIRECTORY {
-				dirRef = child
-				dirRef.HashToBeComputed = true
+
+	for _, subDir := range subDirs {
+		var found bool
+		for _, ref := range dirRef.Children {
+			if ref.Name == subDir && ref.Type == DIRECTORY {
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/reference/object.go:33]: Perhaps we should break immediately if the dirRef.Children do not have the subDir, basically this checking all start from the root path, if any level of the path do not exist in the corresponding dirRef, there must be something wrong.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/reference/object.go (hunk 3 above)
Comment (lpoli) [reply code/go/0chain.net/blobbercore/reference/object.go:33]: It does break with `return` statement.

**Review by peterlimg — APPROVED** [review]
lgtm

**Review by cnlangzi — COMMENTED** [review]
@peterlimg  @lpoli , this PR breaks multiple things
- breaks 204 status for partial delete
- it is incorrect to delete thumbnail if hash is not empty. Because hash always has value even it is empty.

**Pull request merged**